### DISCUSSION TITLE
Adaptive updates and multiple partitions

### DISCUSSION THREAD

**Discussion by sl8vz** [rauc/rauc#1012] [Q&A]
Hi,

Not an issue but more a simple design question.

Before the introduction of adaptive updates, we designed the system with separated main rootfs slot (which doesn't change much) and application fs slot (more frequent updates) and therefore prevent unnecessary updates.

With the introduction of adaptive updates it seems to me that this separation doesn't bring any value anymore and using a single rootfs will more or less lead to the same performance.

If I understand correctly, a given block of data is either skipped, copied from the other bundle or finally downloaded. In this case I don't see the need to manage a separated applications partitions. Am I right ?

Thanks,
Sebastien

**Comment by ejoerns** [comment]
@sl8vz So you had a bundle with rootfs + appfs and used the ability to skip slots with equal hash?

But this only saves installation time but no download size.

Adaptive updates on the other hand target primarily on saving download size instead of preventing unnecessary updates but yes, data that is already at the right place on the block device will not be re-written. To actually not change much of the file system when changing the application, additional care must be taken during image generation.

Maybe something like https://github.com/rauc/rauc/issues/969 could also be of interest for you?

**Comment by sl8vz** [comment]
Thx for your time @ejoerns ,

> @sl8vz So you had a bundle with rootfs + appfs and used the ability to skip slots with equal hash?

It's still in design phase and I just discovered the release of adaptive updates :) Trying to find the most effective solution.

> But this only saves installation time but no download size.

So I misunderstood how the ability to skip slots works. How does that work with HTTP streaming ?
I thought that skipping a slot would also prevent the download but that it would require the whole slot to be unchanged. A single byte change would cause the slot to be fully updated (both download and update) (And this is where adaptive update can help a lot)

> To actually not change much of the file system when changing the application, additional care must be taken during image generation.

This is quite tricky indeed to do in the build process. So it seems that having a separate partition for the apps is still relevant to make sure that the main rootfs is not affected by frequent custom apps updates.

#969 is interesting indeed, we have a similar approach, one appfs per bundle and a single docker partition (but we let docker perform the update itself).

**Reply by ejoerns** [reply]
> Thx for your time @ejoerns ,
> 
> > @sl8vz So you had a bundle with rootfs + appfs and used the ability to skip slots with equal hash?
> 
> It's still in design phase and I just discovered the release of adaptive updates :) Trying to find the most effective solution.
> 
> > But this only saves installation time but no download size.
> 
> So I misunderstood how the ability to skip slots works. How does that work with HTTP streaming ? I thought that skipping a slot would also prevent the download but that it would require the whole slot to be unchanged. A single byte change would cause the slot to be fully updated (both download and update) (And this is where adaptive update can help a lot)

When combining with streaming this is valid, yes, but with the exceptions you noticed. You have to make sure that you place the exact same image inside the bundle that was installed on the system. Then you can benefit from not having to download it if skipping is enabled. This obviously this only works fine if you exactly know which versions you have in the field.

> > To actually not change much of the file system when changing the application, additional care must be taken during image generation.
> 
> This is quite tricky indeed to do in the build process. So it seems that having a separate partition for the apps is still relevant to make sure that the main rootfs is not affected by frequent custom apps updates.

If you just want to save download size and the additional writing time (compared to skipping rootfs write) is not critical, then adaptive method block-hash-index should still be a good option for you.

> #969 is interesting indeed, we have a similar approach, one appfs per bundle and a single docker partition (but we let docker perform the update itself).

Yes, this mainly depends on how you want to deploy applications. If you want to deploy them over access to a central docker registry, then this is fine. If it should be part of the update bundle instead, then the mentioned approach should be an option.

**Reply by sl8vz** [reply]
> If you just want to save download size and the additional writing time (compared to skipping rootfs write) is not critical, then adaptive method block-hash-index should still be a good option for you.

Just to confirm, you mean having separated apps partition + adaptive method, so both mechanisms can be leveraged ?

**Reply by jluebbe** [reply]
> #969 is interesting indeed, we have a similar approach, one appfs per bundle and a single docker partition (but we let docker perform the update itself).

Could you share some details how your current approach works? Knowing some existing solutions should help us with the detailed design for #969.

**Reply by sl8vz** [reply]
> Could you share some details how your current approach works? Knowing some existing solutions should help us with the detailed design for https://github.com/rauc/rauc/issues/969.

Systems partitions are as follow:
- u-boot / u-boot env / u-boot script
- boot.0 (kernel + dtb)
- boot.1
- rootfs.0 (system)
- rootfs.1
- appfs.0 (our custom apps)
- appfs.1
- docker (created on first boot using systemd repart)
- data (expanded using systemd growfs, but is present in the initial sdcard)

Update bundle is (boot, rootfs, appfs)

My original idea was to separated appfs and rootfs and often skip the full rootfs slot. However this is hard to achieve at the build level. (ending-up with the same checksum would require extra logic)

However, having them separate helps to keep the main rootfs binary almost identical and makes the adaptive update more effective.

I was also thinking that when the slots are read-only (which should be the case anyway), the adaptive method could completely skip writing any data if the update slot hash-index is identical to the update bundle hash ? It seems to me that data is always written (not download but copied), right ?
